go/carmen/database_test.go

```diff
@@ -845,6 +845,374 @@ func TestDatabase_CloseDB_Unfinished_Queries(t *testing.T) {
+	for i := 0; i < numBlocks; i++ {
+		for j := 0; j < numAccounts; j++ {
+			addr := Address{byte(j)}
+
+			nonce, complete, err := recovered.GetNonce(roots[i], addr)
+			if err != nil {
+				t.Errorf("cannot get nonce: %v", err)
+			}
+
+			if !complete {
+				t.Errorf("proof is not complete")
+			}
+
+			if got, want := nonce, uint64(i<<8+j+1); got != want {
+				t.Errorf("unexpected nonce, wanted %v, got %v", want, got)
+			}
+
+			/// proof keys
+			for k := 0; k < numKeys; k++ {
+				key := Key{byte(k)}
+				value, complete, err := recovered.GetState(roots[i], addr, key)
+				if err != nil {
+					t.Errorf("cannot get state: %v", err)
+				}
+				if !complete {
+					t.Errorf("proof is not complete")
+				}
+				if got, want := value, (Value{byte(i), byte(j), byte(k)}); got != want {
+					t.Errorf("unexpected value, wanted %v, got %v", want, got)
+				}
+			}
+		}
+	}
+}
+
+func TestDatabase_CloseDB_Unfinished_ProofE_CannotCloseDb(t *testing.T) {
+	db, err := openTestDatabase(t)
+	if err != nil {
+		t.Fatalf("failed to open database: %v", err)
+	}
+
+	if err := db.AddBlock(1, func(context HeadBlockContext) error {
+		if err := context.RunTransaction(func(tc TransactionContext) error {
+			addr := Address{1}
+			tc.CreateAccount(addr)
+			tc.SetNonce(addr, uint64(1))
+			return nil
+		}); err != nil {
+			return err
+		}
+		return nil
+	}); err != nil {
+		t.Fatalf("cannot add block: %v", err)
+	}
+
+	if err := db.Flush(); err != nil {
+		t.Fatalf("cannot flush db: %v", err)
+	}
+
+	block, err := db.GetHistoricContext(1)
+	if err != nil {
+		t.Fatalf("cannot get block: %v", err)
+	}
+
+	var done atomic.Bool
+	go func() {
+		// attempts to close db must keep failing
+		if err := db.Close(); !errors.Is(err, errBlockContextRunning) {
+			t.Errorf("closing db should fail: %v", err)
+		}
+		done.Store(true)
+	}()
+
+	for !done.Load() {
+		_, err = block.GetProof(Address{1})
+		if err != nil {
+			t.Fatalf("cannot get proof: %v", err)
+		}
+	}
+
+	if err := block.Close(); err != nil {
+		t.Fatalf("cannot close block: %v", err)
+	}
+	if err := db.Close(); err != nil {
+		t.Fatalf("cannot close db: %v", err)
+	}
+}
+
+func TestDatabase_ClosedBlock_CannotGetProof(t *testing.T) {
+	db, err := openTestDatabase(t)
+	if err != nil {
+		t.Fatalf("failed to open database: %v", err)
+	}
+
+	if err := db.AddBlock(1, func(context HeadBlockContext) error {
+		if err := context.RunTransaction(func(tc TransactionContext) error {
+			addr := Address{1}
+			tc.CreateAccount(addr)
+			tc.SetNonce(addr, uint64(1))
+			return nil
+		}); err != nil {
+			return err
+		}
+		return nil
+	}); err != nil {
+		t.Fatalf("cannot add block: %v", err)
+	}
+
+	if err := db.Flush(); err != nil {
+		t.Fatalf("cannot flush db: %v", err)
+	}
+
+	block, err := db.GetHistoricContext(1)
+	if err != nil {
+		t.Fatalf("cannot get block: %v", err)
+	}
+
+	if err := block.Close(); err != nil {
+		t.Fatalf("cannot close block: %v", err)
+	}
+
+	if err := db.Close(); err != nil {
```

Comment:
I think the `db.Close()` should be moved after the attempt to fetch a proof since the closing of the block context needs to prevent further proof reading, not the closing of the DB.

Reply:
right, moved